synapse/handlers/oidc_handler.py

```diff
@@ -0,0 +1,659 @@
+
+        return self._provider_metadata
+
+    async def load_jwks(self, force=False) -> Dict[str, List[Dict[str, str]]]:
+        """Load the JSON Web Key Set used to sign ID tokens.
+
+        If we're not using the ``userinfo_endpoint``, user infos are extracted
+        from the ID token, which is a JWT signed by keys given by the provider.
+        The keys are then cached.
+
+        Args:
+            force (bool): Force reloading the keys.
+
+        Returns:
+            dict: The key set
+
+            Looks like this::
+
+                {
+                    'keys': [
+                        {
+                            'kid': 'abcdef',
+                            'kty': 'RSA',
+                            'alg': 'RS256',
+                            'use': 'sig',
+                            'e': 'XXXX',
+                            'n': 'XXXX',
+                        }
+                    ]
+                }
+        """
+        if self._uses_userinfo:
+            # We're not using jwt signing, return an empty jwk set
+            return {"keys": []}
+
+        metadata = await self.load_metadata()
+        jwk_set = metadata.get("jwks")
+        if jwk_set is not None and not force:
+            return jwk_set
+
+        uri = metadata.get("jwks_uri")
+        if not uri:
+            raise RuntimeError('Missing "jwks_uri" in metadata')
+
+        jwk_set = await self._http_client.get_json(uri)
+        self._provider_metadata["jwks"] = jwk_set
+        return jwk_set
+
+    async def _exchange_code(self, code: str) -> Dict[str, str]:
+        """Exchange an authorization code for a token.
+
+        This calls the ``token_endpoint`` with the authorization code we
+        received in the callback to exchange it for a token. The call uses the
+        ``ClientAuth`` to authenticate with the client with its ID and secret.
+
+        Args:
+            code (str): The autorization code we got from the callback.
+
+        Returns:
+            dict: contains various tokens.
+
+            May look like this::
+
+                {
+                    'token_type': 'bearer',
+                    'access_token': 'abcdef',
+                    'expires_in': 3599,
+                    'id_token': 'ghijkl',
+                    'refresh_token': 'mnopqr',
+                }
+
+        Raises:
+            OidcError: when the ``token_endpoint`` returned an error.
+        """
+        metadata = await self.load_metadata()
+        token_endpoint = metadata.get("token_endpoint")
+        args = {
+            "grant_type": "authorization_code",
+            "code": code,
+            "redirect_uri": self._callback_url,
+        }
+
+        uri, headers, body = self._client_auth.prepare(
+            method="POST", uri=token_endpoint, headers={}, body=""
+        )
+        headers = {k: [v] for (k, v) in headers.items()}
+        qs = parse_qsl(body, keep_blank_values=True)
+        args.update(qs)
```

Comment:
This is converting from form-encoded data to arguments? How come this is necessary? Is it just to use `post_urlencoded_get_json`?

Reply:
ClientAuth.prepare from authlib will fill the body/headers depending on what auth method is used (`client_secret_post` fills the body, `client_secret_post` uses the `Authorization` header). Unfortunately, the body is returned as a string, not a dict.

Maybe I should pass the args to `prepare` and use `_http_client.request` instead.

Reply:
And that's what I ended up doing. It makes testing a bit harder, but I think the code exchange part is a bit clearer now.